google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/SeekableByteChannelPrefetcher.java

```diff
@@ -0,0 +1,494 @@
+      // |---:--position->
+      //     :<--offset-->
+      //     ^ blockIndex*bufSize
+      int availableToCopy = src.position() - offset;
+      if (availableToCopy < 0) {
+        // the caller is asking to read past the end of the file
+        nbReadsPastEnd++;
+        return -1; // EOF
+      }
+      int bytesToCopy = dst.remaining();
+      byte[] array = src.array();
+      if (availableToCopy < bytesToCopy) {
+        bytesToCopy = availableToCopy;
+      }
+      dst.put(array, offset, bytesToCopy);
+      position += bytesToCopy;
+      if (trackTime) {
+        msCopyingData += copyingData.elapsed(TimeUnit.MILLISECONDS);
+      }
+      bytesReturned += bytesToCopy;
+      if (availableToCopy == 0) {
+        // EOF
+        return -1;
+      }
+      return bytesToCopy;
+    } finally {
+      if (trackTime) {
+        betweenCallsToRead.reset();
+        betweenCallsToRead.start();
+      }
+    }
+  }
+
+  /**
+   * Writing isn't supported.
+   */
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Returns this channel's position.
+   *
+   * @return This channel's position,
+   * a non-negative integer counting the number of bytes
+   * from the beginning of the entity to the current position
+   * @throws ClosedChannelException If this channel is closed
+   * @throws IOException            If some other I/O error occurs
+   */
+  @Override
+  public long position() throws IOException {
+    if (!open) throw new ClosedChannelException();
+    return position;
+  }
+
+  /**
+   * Sets this channel's position.
+   * <p>
+   * <p> Setting the position to a value that is greater than the current size
+   * is legal but does not change the size of the entity.  A later attempt to
+   * read bytes at such a position will immediately return an end-of-file
+   * indication.  A later attempt to write bytes at such a position will cause
+   * the entity to grow to accommodate the new bytes; the values of any bytes
+   * between the previous end-of-file and the newly-written bytes are
+   * unspecified.
+   * <p>
+   * <p> Setting the channel's position is not recommended when connected to
+   * an entity, typically a file, that is opened with the {@link
+   * java.nio.file.StandardOpenOption#APPEND APPEND} option. When opened for
+   * append, the position is first advanced to the end before writing.
+   *
+   * @param newPosition The new position, a non-negative integer counting
+   *                    the number of bytes from the beginning of the entity
+   * @return This channel
+   * @throws ClosedChannelException   If this channel is closed
+   * @throws IllegalArgumentException If the new position is negative
+   * @throws IOException              If some other I/O error occurs
+   */
+  @Override
+  public SeekableByteChannel position(long newPosition) throws IOException {
+    if (!open) throw new ClosedChannelException();
```

Comment:
The design is that we start fetching on the first read. If the user is going to read at this position then they typically issue a read almost immediately, negating the benefit from this change. If the user code is not going to read (it's issuing a series of seeks because that's how the code was written) then this change would incur a large penalty.

So this would be taking a large risk for not enough benefit to be worthwhile, in my opinion.